[move-only] A group of batched changes

gottesmm opened the pull request:
1. First change converts __shared -> borrowed in the tests. I did this since semantically there isn't a difference beyond mangling and it makes sense to match what our users will use.
2. I turned off emitting move only errors on functions if we already emitted an escaping capture bug for that function. The reason why is because SILGen today assumes that we will error in such cases and thus does not emit markers in said function for the inout. This then causes us to emit spurious "found a copy of a noncopyable value" errors that
may cause the user to think there is a bug in the compiler.

gottesmm committed:
[move-only] If we diagnose an invalid escaping capture, turn off noncopyable checking in the closure.

The reason why I am doing this is that currently SILGen knows when emitting said
closure that we are going to emit an error (that it does not have enough
information to emit itself since it doesn't know the caller), so doesn't emit
move checking markers. The result is that the move checker will not eliminate
any copies in the closure and thus will emit a "copy of noncopyable type found"
error and tell the user to file a bug. This just suppresses that.

rdar://108511866

gottesmm committed:
[move-only] When emitting borrows for move only types, use a load [copy] instead of a load_borrow.

The reason why I am doing this is that otherwise if one has a function that
takes both a guaranteed and an owned parameter, we will break OSSA invariants
since the load [take] will invalidate the load_borrow. So instead, we put in a
load_borrow knowing that the move checker will convert it to a load_borrow
assuming that the two pass exclusivity checking.

NOTE: Because of some missing functionality in subsequent tests, I had to
disable one test (moveonly_escaping_definite_initialization.swift) and also add
some checks for copy of noncopyable object errors. They will go away in the next
2 commits.

rdar://108510987

gottesmm commented:
@swift-ci test

gottesmm commented:
In a forthcoming commit, the 2nd commit will come into play, I am just peeling the onion

gottesmm commented:
@swift-ci smoke test

kavon commented on include/swift/AST/SemanticAttrs.def:
```diff
@@ -127,8 +127,21 @@ SEMANTICS_ATTR(LIFETIMEMANAGEMENT_COPY, "lifetimemanagement.copy")
+// errors twice, once on the specialized and once on the original
+// function. Instead, we put this on the original function.
+//
+// 2. If we emit a diagnose invalid escaping captures error due to an inout
+// being escaped into an escaping closure, we do not want to emit move errors in
+// the closure. This is because SILGen today assumes that we will error in such
+// cases and thus does not emit markers in said function for the inout. This
+// then causes us to emit spurious "found a copy of a noncopyable value" errors
+// that may cause the user to think there is a bug in the compiler.
 SEMANTICS_ATTR(NO_MOVEONLY_DIAGNOSTICS, "sil.optimizer.moveonly.diagnostic.ignore")
```
Can you add a regression test for this?

gottesmm replied:
I'm going to add the regression test in a follow on commit. I am adding in a follow on commit a large suite of closure tests. The reason why I am not including it here is that it crashes on main due to a different issue that I am going to fix.

gottesmm committed:
[move-only] Instead of using AccessUseVisitor to visit addresses use TransitiveAddressVisitor.

This makes it so that the move address checker is not dependent on starting the
traversal at a base object. I also included verifier checks that the API can
visit all address uses for:

1. project_box.
2. alloc_stack.
3. ref_element_addr.
4. ref_tail_addr.
5. global_addr_inst.

this is because this visitor is now apart of the SIL API definition as being
able to enumerate /all/ addresses derived from a specific chosen address value.

This is a refactoring NFCI change.

rdar://108510644

gottesmm committed:
[generic-spec] Add an option -sil-generic-dump-functions-after-specialization that makes it easy to look for errors around specialization.

gottesmm committed:
[move-only] Eliminate dead code that was not removed when re-projected move only types were removed.

gottesmm committed:
[move-only] When emitting accesses to let boxes containing a noncopyable type, always emit mark_must_check.

The reason to do this is that:

1. Otherwise, we do not emit markers when someone attempts to consume the let.
We need the no_consume_or_assign to be there.
2. We need to insert assign_but_not_consuming so that DI can properly check lets
that are conditionally initialized and convert them to
initable_but_not_consuming.

I included a full definite_init SIL test that validates that we get the correct
codegen after DI in this case and emit the appropriate error as well.

rdar://108511534

gottesmm committed:
[move-only] Add a large SILGen/-verify test that checks captures of escaping closure behavior.

NOTE: We consider consuming closures that are marked without escaping as
non-escaping. We shouldn't do this, but for now I just included them so that
their behavior is in some test and we at least don't crash.

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test source compatibility

gottesmm commented:
@swift-ci test

gottesmm committed:
[move-only] Clone all __owned tests into consuming tests.

I left in the __owned tests of course.

rdar://108511703

atrick commented on lib/SIL/Utils/OwnershipUtils.cpp:
```diff
@@ -1137,56 +1144,73 @@ swift::findTransitiveUsesForAddress(SILValue projectedAddress,
-        } else {
-          llvm::copy(beginApply->getTokenResult()->getUses(),
-                     std::back_inserter(*foundUses));
+      // TODO: the empty check should not be needed when dead begin_apply is
```
Why are `begin_apply` and `load_borrow` handled in different places? They seem like equivalent problems.

gottesmm replied:
You are right. I can do a similar treatment to what I did with load_borrow.

gottesmm replied:
If I change this, I want to do it in a subsequent commit that isn't so big.

gottesmm replied:
I took care of it since I needed to do some more work anyways.

atrick commented on lib/SIL/Utils/OwnershipUtils.cpp:
```diff
@@ -1118,15 +1118,22 @@ swift::findTransitiveUsesForAddress(SILValue projectedAddress,
+        isa<UncheckedRefCastAddrInst>(user) || isa<KeyPathInst>(user) ||
+        isa<RetainValueAddrInst>(user) || isa<ReleaseValueAddrInst>(user) ||
+        isa<PackElementSetInst>(user) || isa<PackElementGetInst>(user) ||
+        isa<DeinitExistentialAddrInst>(user) || isa<LoadBorrowInst>(user)) {
```
You may want to check the opcode list against `swift::visitAccessedAddress`

gottesmm replied:
I removed the comment.

atrick reviewed (approved):
I only reviewed the commit "[move-only] Instead of using AccessUseVisitor to visit addresses"

Looks good. Just a few comments inline.

Ideally the starting address producers would be in sync with `AccessUseDefChainVisitor<Impl, Result>::visit`. We're just missing a few of the unusual cases.

gottesmm committed:
[move-only] When emitting exclusivity diagnostics for move only types, do not suggest to the user to make a local copy.

It doesn't make sense to give this note since one can't make a copy of a
noncopyable type.

rdar://108511627

gottesmm committed:
[move-only] Convert __shared to borrowing in move only tests.

These are the same semantically, just the mangling is slightly different. The
benefit of doing this is that we are actually testing what we expect our users
to do.

rdar://108511703

kavon commented on test/SILGen/moveonly_escaping_closure.swift:
```diff
@@ -0,0 +1,1583 @@
+// CHECK: end_access [[ACCESS]]
+// CHECK: } // end sil function '$s16moveonly_closure026testInOutVarClosureCaptureF0yyyyczF'
+//
+// CHECK-LABEL: sil private [ossa] @$s16moveonly_closure026testInOutVarClosureCaptureF0yyyyczFyycfU_ : $@convention(thin) (@guaranteed { var SingleElt }) -> () {
+// CHECK: bb0([[BOX:%.*]] :
+// CHECK:   [[PROJECT:%.*]] = project_box [[BOX]]
```
Most of the SILGen for the closure bodies appear to have exactly the same pattern (since they have the same bodies), the only difference being whether there's a `project_box` or in one case, no projection because it takes an `(@inout_aliasable SingleElt)`, for the captured var. 

Would it be reasonable to cut down on amount of CHECK's for bits that are already covered by another test, so it's only checking the important parts?

For example, my guess is that this is ensuring there is a `mark_must_check` and then a `load [copy]` and no load_borrow's? If it's possible to cut down on the redundancy, it would make it easier to grok what's important (e.g., even dropping CHECK's for destroy's and end_accesses since they'd be caught by verifier anyway).

gottesmm replied:
I am purposely trying to be verbose here so that if someone changes the codegen for any of these individual parts, they need to update this codegen and think about what they are doing. So making the pattern matching looser would go against specifically what I am trying to do.

kavon commented on test/SILGen/moveonly_escaping_closure.swift:
```diff
@@ -0,0 +1,1583 @@
+// MARK: Let Capture Tests //
+/////////////////////////////
+
+// NOTE: In the let tests below, we often times have multiple mark_must_check on
+// the same project_box. If we just looked at the SILGen output, the move
+// checker would potentially not work correctly upon that code and would emit
+// weird diagnostics since the AccessVisitor will visit /all/ uses on all of the
+// mark_must_check when we process each individual mark must check. Luckily this
+// is not actually an issue since when allocbox to stack runs, we specialize the
```
The var-capture tests have multiple `mark_must_check`'s. Is it OK for vars but not for lets? If so, maybe worth noting that too. Also, would this explainer make more sense elsewhere in the source code rather than a test, for discoverability?

gottesmm replied:
This comment actually needs to be eliminated since we do not use the AccessVisitor anymore.

kavon commented on test/SILOptimizer/move_only_checker_addressonly_fail.swift:
```diff
@@ -12,6 +12,8 @@ func test1<T>(_ x: T) {
 
     // We are ok with the checker not supporting this today as long as we emit
     // the error msg above.
-    consumeValue(x2) // expected-error {{Usage of @noImplicitCopy that the move checker does not know how to check!}}
-    consumeValue(x2) // expected-error {{Usage of @noImplicitCopy that the move checker does not know how to check!}}
+    consumeValue(x2) // expected-error {{'x2' has guaranteed ownership but was consumed}}
+    // expected-note @-1 {{consuming use here}}
+    consumeValue(x2) // expected-error {{'x2' has guaranteed ownership but was consumed}}
+    // expected-note @-1 {{consuming use here}}
```
Nice! Guess we can update that comment since this is now supported 😄

kavon reviewed (commented):
Looks fine overall. Just some API suggestions and questions.

gottesmm commented on test/SIL/verifier-fail-address-phi-ossa.sil:
```diff
@@ -1,7 +1,9 @@
 // RUN: %empty-directory(%t)
-// RUN: not --crash %target-sil-opt -enable-sil-verify-all %s 2> %t/err.txt
+// RUN: not --crash %target-sil-opt -sil-verifier-disable-transitive-uses-for-address-walker-check -enable-sil-verify-all %s 2> %t/err.txt
```
The issue is that both verifiers crash on the same behavior. I am disabling the transitive address walker verifier so this test still only hits the original verification.

kavon replied:
If that's the case, we should duplicate the RUN line so it crashes on the original verification (with the disable flag), and then another where it crashes on the other verifier (without the disable flag).

gottesmm replied:
Then I would have to add another disable option for the other verifier. I don't think this is really worth the complexity.

gottesmm commented on include/swift/SIL/OwnershipUtils.h:
```diff
@@ -769,13 +770,97 @@ inline AddressUseKind meet(AddressUseKind lhs, AddressUseKind rhs) {
+public:
+  virtual ~TransitiveAddressVisitor() {}
+
+  /// Customization point for visiting uses. Returns true if we should continue
+  /// visiting.
+  ///
+  /// NOTE: Do not call this directly from within
+  /// findTransitiveUsesForAddress. Please call callVisitUse. This is intended
+  /// just for subclasses to override.
+  virtual bool visitUse(Operand *use) { return true; }
```
I specifically left the customization point as visitUse so that when users implement this, they override visitUse.

kavon replied:
I think you misunderstood me. `callVisitUse` and `visitUse` are both `public`, yet one has a warning saying it's only intended to be overridden by subclasses. So why is the virtual method not `protected`?

gottesmm replied:
I can make it protected.

gottesmm commented on lib/SIL/Verifier/SILVerifier.cpp:
```diff
@@ -702,6 +702,32 @@ struct ImmutableAddressUseVerifier {
+static llvm::cl::opt<bool> DisableTransitiveUsesForAddressWalkerChecker(
+    "sil-verifier-disable-transitive-uses-for-address-walker-check",
+    llvm::cl::init(false),
+    llvm::cl::desc("Certain tests want to validate that certain verification "
+                   "checks work that the address walker to check also errors "
+                   "on. This turns off this checker for testing purposes. "
+                   "E.x.: Testing that address phis are rejected in OSSA"));
```
I can word smith this a bit.

gottesmm commented on include/swift/SIL/OwnershipUtils.h:
```diff
@@ -769,13 +770,97 @@ inline AddressUseKind meet(AddressUseKind lhs, AddressUseKind rhs) {
   return (lhs > rhs) ? lhs : rhs;
 }
 
+/// A state structure for findTransitiveUsesForAddress. Intended to be only used
+/// a single time. Please always use a new one for every call to
+/// findTransitiveUsesForAddress.
```
I don't want to introduce a unique_ptr or a moveable rvalue method (mainly b/c even though we use that in parts of the code base, it is a foreign feature to many users).

I think a better pattern would be to make the result Optional and ban the copy constructor. Then when findTransitiveUsesForAddress completes, we invalidate the result. Then if someone tries to call it again, they get a run time crash.

kavon replied:
A runtime crash is not better than a static check by the compiler that the API is used correctly. The only users of this API are Swift compiler developers, who have to understand movable rvalue methods as we use it elsewhere.

gottesmm replied:
@kavon I disagree. If someone misuses this API and get a runtime crash, it is going to happen immediately when they use the API... that suggests either way someone will get a quick slap on the wrist showing them what the issue is. I think this is a matter of personal style and taste.

kavon replied:
> If someone misuses this API and get a runtime crash, it is going to happen immediately when they use the API

By the nature of runtime tests, this is not guaranteed. The following gets statically rejected by an rvalue-method version, but may not be caught by testing and would be a subtle bug:

```
MyVisitor visitor;
findTransitiveUsesForAddress(address, visitor);
// ...
if (uncommonScenario)
  findTransitiveUsesForAddress(address, visitor);
```

This is not a matter of taste. Relying on a runtime test to enforce lifetime rules is strictly less robust and mistake-prone. Please make use of C++'s lifetime features as we do elsewhere in the compiler to prevent mistakes.

gottesmm committed:
Update test for changes in https://github.com/apple/swift/pull/65377

That PR was for rdar://108385761

gottesmm commented:
@swift-ci test source compatibility

gottesmm commented:
I discovered a small issue where I introduced an optnone by mistake. I am going to remove it in a follow on smoke test change.

gottesmm commented:
Actually I was wrong about the optnone. It was in my local branch that I am working with on top of this.

gottesmm commented:
I took a look at the compatibility failure. One was a UPASS that showed up on the actual bot.

gottesmm commented:
The other failure is distributed actors which is also failing on other PRs due to misconfiguration. I talked with Konrad and he agreed I can ignore it.
